File: ghosts.py
import pygame
import random
from entities import Entity
from settings import *

vec = pygame.math.Vector2

class Ghost(Entity):
    def __init__(self, x, y, color_key: str):
        super().__init__(x, y)
        self.color = COLORS[color_key]
        self.direction = vec(1, 0)

    def update(self):
        if int(self.pix_pos.x) % TILE_SIZE == 0 and int(self.pix_pos.y) % TILE_SIZE == 0:
            self.grid_pos = self.pix_pos // TILE_SIZE
            self.direction = self.calculate_next_move()
        
        self.pix_pos += self.direction * self.speed

    def calculate_next_move(self) -> vec:
        directions = [vec(0,1), vec(0,-1), vec(1,0), vec(-1,0)]
        # Impede que o fantasma volte pelo caminho que veio (comportamento clássico)
        valid_moves = [d for d in directions if self.can_move(d) and d != -self.direction]
        
        if not valid_moves: # Bateu num beco sem saída
            return -self.direction
        return random.choice(valid_moves)

    def draw(self, screen: pygame.Surface):
        rect = (int(self.pix_pos.x + 4), int(self.pix_pos.y + 4), TILE_SIZE - 8, TILE_SIZE - 8)
        pygame.draw.rect(screen, self.color, rect, border_radius=6)

File: ghosts.js
import { CONFIG, COLORS, MAP_DATA } from './settings.js';
import { Entity } from './entities.js';

export class Ghost extends Entity {
    constructor(x, y, color, targetCanto) {
        super(x, y);
        this.color = color;
        this.mode = 'CHASE'; // SCATTER, CHASE, FRIGHTENED
        this.timer = 0;
        this.targetCanto = targetCanto;
        this.velocity = 1.8;
    }

    update(pacmanPos) {
        this.timer++;
        
        // Máquina de Estado Simples (Troca a cada 20 segundos)
        if (this.timer > 1200) {
            this.mode = this.mode === 'CHASE' ? 'SCATTER' : 'CHASE';
            this.timer = 0;
        }

        if (this.isAtCenter()) {
            this.gridPos = { x: this.position.x / CONFIG.TILE_SIZE, y: this.position.y / CONFIG.TILE_SIZE };
            const target = this.mode === 'CHASE' ? pacmanPos : this.targetCanto;
            this.direction = this.getBestMove(target);
        }

        this.position.x += this.direction.x * this.velocity;
        this.position.y += this.direction.y * this.velocity;
    }

    getBestMove(target) {
        const moves = this.getPossibleMoves();
        // IA não pode voltar para trás
        const filtered = moves.filter(m => m.x !== -this.direction.x || m.y !== -this.direction.y);
        
        let bestMove = filtered[0] || moves[0];
        let minDist = Infinity;

        for (const move of filtered) {
            const nextX = this.gridPos.x + move.x;
            const nextY = this.gridPos.y + move.y;
            const dist = Math.hypot(nextX - target.x, nextY - target.y);
            if (dist < minDist) {
                minDist = dist;
                bestMove = move;
            }
        }
        return bestMove;
    }

    draw(ctx) {
        const x = this.position.x;
        const y = this.position.y;
        
        ctx.fillStyle = this.color;
        // Cabeça
        ctx.beginPath();
        ctx.arc(x + 16, y + 16, 14, Math.PI, 0);
        ctx.lineTo(x + 30, y + 30);
        ctx.lineTo(x + 2, y + 30);
        ctx.fill();
        
        // Olhos
        ctx.fillStyle = "white";
        ctx.beginPath();
        ctx.arc(x + 10, y + 12, 4, 0, Math.PI * 2);
        ctx.arc(x + 22, y + 12, 4, 0, Math.PI * 2);
        ctx.fill();
    }
}
// ... (Aqui entrariam mais 350 linhas detalhando 4 classes 
// diferentes: Blinky, Pinky, Inky e Clyde, cada um com sua estratégia única)
